File: paddler_client_python/paddler_client/grammar_constraint.py
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field


class GbnfGrammarConstraint(BaseModel):
    type: Literal["gbnf"] = "gbnf"
    grammar: str
    root: str


class JsonSchemaGrammarConstraint(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    type: Literal["json_schema"] = "json_schema"
    schema_value: Annotated[str, Field(alias="schema")]


GrammarConstraint = GbnfGrammarConstraint | JsonSchemaGrammarConstraint

File: paddler_client_javascript/src/schemas/GrammarConstraint.ts
import { z } from "zod";

export const GrammarConstraintSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("gbnf"),
    grammar: z.string(),
    root: z.string(),
  }),
  z.object({
    type: z.literal("json_schema"),
    schema: z.string(),
  }),
]);

export type GrammarConstraint = z.infer<typeof GrammarConstraintSchema>;
